### PR TITLE
t2 run/push: explicit compression

### PR DESCRIPTION
**Pull request by rwaldron**
- Attempt to use acorn.parse first
- Fallback to uglify.parse
- explicitly define all parser, compressor, and mangling options
  - improves compression

### CODE REVIEW

**Comment by rwaldron** [comment]
On hold until I figure out why sometimes builds just don't happen...

**Comment by rwaldron** [comment]
The issue I was hitting was literally caused by a `process.exit()` I had added after pushing and during test ing of something completely unrelated. WHOOPS


@johnnyman727 review?

File: lib/tessel/deploy.js [hunk 1]
```diff
@@ -832,30 +833,72 @@ actions.project = function(options) {
...
+    // (but unfortunately, not perfect)
+    ast = uglify.AST_Node.from_mozilla_ast(
+      acorn.parse(source, {
+        allowImportExportEverywhere: true,
+        allowReturnOutsideFunction: true,
+        ecmaVersion: 7,
+      })
+    );
+    // However, if it fails, also try uglify...
+  } catch (error) {
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: Are there any kinds of errors returned from Acorn that might indicate that we should not continue?

File: lib/tessel/deploy.js [hunk 2]
```diff
@@ -832,30 +833,72 @@ actions.project = function(options) {
...
+    conditionals: true,
+    comparisons: true,
+    evaluate: true,
+    hoist_funs: true,
+    hoist_vars: true,
+    if_return: true,
+    join_vars: true,
+    loops: true,
+    properties: true,
+    screw_ie8: true,
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: hahaha

File: lib/tessel/deploy.js [hunk 3]
```diff
@@ -3,6 +3,7 @@ var path = require('path');
 var zlib = require('zlib');
 
 // Third Party Dependencies
+var acorn = require('acorn');
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: I do wonder how much time this adds before the CLI responds to a command :/

**Comment by johnnyman727** [comment]
@rwaldron looks okay to me. Do we have a metric for how much better the compression is?

File: lib/tessel/deploy.js (hunk 3 above)
Comment (rwaldron) [inline lib/tessel/deploy.js]: I will do some profiling.

Comment (rwaldron) [inline lib/tessel/deploy.js]: FWIW, `acorn.parse` is significantly faster than the built-in uglifyjs2 parser ([by the author's own admission](http://lisperator.net/blog/uglifyjs-why-not-switching-to-spidermonkey-ast/)). Ideally, code will only have to go through `acorn.parse`, so it _should_ improve performance. Either way, I will put together some profiles.

**Comment by rwaldron** [comment]
@johnnyman727 with the same method that I used [previously](https://github.com/tessel/t2-cli/issues/490#issuecomment-181524393) (ie. same patch). 


Here are the first 10 samples of each test, which is _only_ measuring the change to the time it takes to resolve all `require()s`: 


| Parser/Compressor | Seconds |
| -------- | ------- |
| uglify | 1.448241 |
| uglify | 0.758325 |
| uglify | 0.784999 |
| uglify | 0.791185 |
| uglify | 0.780991 |
| uglify | 0.773770 |
| uglify | 0.788832 |
| uglify | 0.765376 |
| uglify | 0.768271 |
| uglify | 0.811554 |
| acorn+uglify | 1.032818 |
| acorn+uglify | 0.784385 |
| acorn+uglify | 0.813502 |
| acorn+uglify | 0.782333 |
| acorn+uglify | 0.757092 |
| acorn+uglify | 0.766412 |
| acorn+uglify | 0.778228 |
| acorn+uglify | 0.774978 |
| acorn+uglify | 0.792304 |
| acorn+uglify | 0.800180 |

(Raw data available here: https://dl.dropboxusercontent.com/u/3531958/2016-02-16-require-resolution-time.txt)

Next, I measured the time it takes from when all deps have been resolved (but just before compression), to just after compression. This patch will reproduce my measurement approach: 


```patch
From 2a33231ea1ca400d4e8ac2c1c6441ef5984c233e Mon Sep 17 00:00:00 2001
From: Rick Waldron <waldron.rick@gmail.com>
Date: Tue, 16 Feb 2016 11:51:47 -0500
Subject: [PATCH] measure

---
 lib/tessel/deploy.js | 32 ++++++++++++++++++++++++++++++--
 1 file changed, 30 insertions(+), 2 deletions(-)

diff --git a/lib/tessel/deploy.js b/lib/tessel/deploy.js
index bb956cb..fb48857 100644
--- a/lib/tessel/deploy.js
+++ b/lib/tessel/deploy.js
@@ -742,6 +742,11 @@ actions.tarBundle = function(opts) {
           var absRoot = path.resolve(globRoot);
           var written = {};
 
+var time = process.hrtime();
+var files = 0;
+var totalByteLength = 0;
+
+
           // 1. Move all dependency entries to the temp directory
           dependencies.forEach(dependency => {
             var isJS = dependency.file.endsWith('.js');
@@ -755,16 +760,39 @@ actions.tarBundle = function(opts) {
             if (isJS) {
               try {
                 source = actions.compress(source);
+                files++;
               } catch (error) {
                 reject(error);
               }
             }
 
-            fs.outputFileSync(target, source);
+            // fs.outputFileSync(target, source);
 
-            written[target] = true;
+            written[target] = source;
           });
 
+
+var diff = process.hrtime(time);
+var ns = (diff[0] * 1e9 + diff[1]);
+var ms = ns / 1e6;
+var ss = ns / 1e9;
+
+console.log("Compressed in: ");
+console.log(`${ns} nanoseconds`);
+console.log(`${ms} milliseconds`);
+console.log(`${ss} seconds`);
+
+var totalByteLength = Object.keys(written).reduce((accum, key) => {
+  return accum + Buffer.byteLength(written[key], 'utf8');
+}, 0);
+
+var data = `"${files} files, ${totalByteLength} bytes",${ss.toFixed(6)}\n`;
+
+require('fs').appendFileSync(path.join(process.cwd(), 'compression-time.txt'), data);
+
+
+process.exit();
+
           // 2. Copy any files that matched all .tesselinclude patterns
           includeFiles.forEach(file => {
             var target = path.join(tempBundleDir, file);
-- 
2.2.1
```




Here is an abbreviated look at the project being deployed: 

```
$ tree -L 3
.
├── README.md
├── index.js
├── node_modules
│   ├── express
│   │   ├── History.md
│   │   ├── LICENSE
│   │   ├── Readme.md
│   │   ├── index.js
│   │   ├── lib
│   │   ├── node_modules
│   │   └── package.json
│   └── socket.io
│       ├── History.md
│       ├── LICENSE
│       ├── Makefile
│       ├── Readme.md
│       ├── index.js
│       ├── lib
│       ├── node_modules
│       └── package.json
├── package.json
└── public
    ├── index.html
    ├── jquery-1.10.2.min.js
    ├── main.js
    └── style.css
```



###  Acorn Parser + UglifyJS w/ explicit compression settings

| # Files | Resulting Size (bytes) | Seconds |
| --------- | ------------ | -------- |
| 163 | 657591 | 3.304039 |
| 163 | 657591 | 3.206970 |
| 163 | 657591 | 2.990535 |
| 163 | 657591 | 3.005475 |
| 163 | 657591 | 2.967336 |
| 163 | 657591 | 3.214797 |
| 163 | 657591 | 3.348803 |
| 163 | 657591 | 3.020216 |
| 163 | 657591 | 3.321714 |
| 163 | 657591 | 3.314362 |

Avg. time = 3.1694247s


###  UglifyJS parser + UglifyJS w/ explicit compression settings

| # Files | Resulting Size (bytes) | Seconds |
| --------- | ------------ | -------- |
| 163 | 657591 | 3.077363 |
| 163 | 657591 | 3.331193 |
| 163 | 657591 | 3.357690 |
| 163 | 657591 | 5.354037 |
| 163 | 657591 | 3.349065 |
| 163 | 657591 | 3.408139 |
| 163 | 657591 | 3.355340 |
| 163 | 657591 | 3.073677 |
| 163 | 657591 | 3.110259 |
| 163 | 657591 | 3.153992 |

Avg. time = 3.4570755s

###  Master

| # Files | Resulting Size (bytes) | Seconds |
| --------- | ------------ | -------- |
| 163 | 744568 | 2.364435 |
| 163 | 744568 | 2.349112 |
| 163 | 744568 | 2.260972 |
| 163 | 744568 | 2.311032 |
| 163 | 744568 | 2.376820 |
| 163 | 744568 | 2.300073 |
| 163 | 744568 | 2.381218 |
| 163 | 744568 | 2.308895 |
| 163 | 744568 | 2.310817 |
| 163 | 744568 | 2.360243 |


Avg. time = 2.3323617s


Avg. time differences:

| Strategy | Avg. Time Cost of Explicit Compression |
| ----------- | -------------------------- |
| acorn + uglify | 0.8370630000000001 |
| uglify | 1.1247137999999999 |


Avg. Aggregate Time Cost of Explicit Compression = 0.9808884s

(Raw data available here: https://dl.dropboxusercontent.com/u/3531958/2016-08-16-compression-time.txt)


-------------------------------


In summary, the branch with explicit compression settings was able to eliminate `86.98Kb`, with an average increase of ~1 second. There was no significant cost to resolving dependencies.

**Comment by johnnyman727** [comment]
@rwaldron thanks for the thorough analysis! Looks like it's a good idea to merge.
